Source/Core/VideoBackends/D3D/PostProcessing.cpp

```diff
@@ -0,0 +1,850 @@
+
+		// Last shader in the list?
+		if (shader_index == (m_post_processing_shaders.size() - 1))
+		{
+			// Last shader in the list, so write to the output texture.
+			shader->Draw(this, visible_rect, tex_size, real_texture, buffer_rect, buffer_size, input_color_texture, m_depth_copy_texture, -1);
+		}
+		else
+		{
+			// To save a copy, we use the output of one shader as the input to the next.
+			// This works except when the last pass is scaled, as the next shader expects a full-size input, so re-use the copy buffer for this case.
+			D3DTexture2D* output_color_texture = (shader->IsLastPassScaled()) ? input_color_texture : shader->GetLastPassOutputTexture();
+			shader->Draw(this, buffer_rect, buffer_size, output_color_texture, buffer_rect, buffer_size, input_color_texture, m_depth_copy_texture, -1);
+			input_color_texture = output_color_texture;
+		}
+	}
+}
+
+bool D3DPostProcessor::ResizeCopyBuffers(const TargetSize& size, int layers)
+{
+	HRESULT hr;
+	if (m_copy_size == size && m_copy_layers == layers)
+		return true;
+
+	// reset before creating, in case it fails
+	SAFE_RELEASE(m_color_copy_texture);
+	SAFE_RELEASE(m_depth_copy_texture);
+	m_copy_size.Set(0, 0);
+	m_copy_layers = 0;
+
+	Microsoft::WRL::ComPtr<ID3D11Texture2D> color_texture;
+	CD3D11_TEXTURE2D_DESC color_desc(DXGI_FORMAT_R8G8B8A8_UNORM, size.width, size.height, layers, 1, D3D11_BIND_SHADER_RESOURCE | D3D11_BIND_RENDER_TARGET, D3D11_USAGE_DEFAULT, 0, 1, 0, 0);
+	if (FAILED(hr = D3D::device->CreateTexture2D(&color_desc, nullptr, &color_texture)))
+	{
+		ERROR_LOG(VIDEO, "CreateTexture2D failed, hr=0x%X", hr);
+		return false;
+	}
+
+	Microsoft::WRL::ComPtr<ID3D11Texture2D> depth_texture;
+	CD3D11_TEXTURE2D_DESC depth_desc(DXGI_FORMAT_R32_FLOAT, size.width, size.height, layers, 1, D3D11_BIND_SHADER_RESOURCE | D3D11_BIND_RENDER_TARGET, D3D11_USAGE_DEFAULT, 0, 1, 0, 0);
+	if (FAILED(hr = D3D::device->CreateTexture2D(&depth_desc, nullptr, &depth_texture)))
+	{
+		ERROR_LOG(VIDEO, "CreateTexture2D failed, hr=0x%X", hr);
+		return false;
+	}
+
+	m_copy_size = size;
+	m_copy_layers = layers;
+	m_color_copy_texture = new D3DTexture2D(color_texture.Get(), (D3D11_BIND_FLAG)color_desc.BindFlags, color_desc.Format);
+	m_depth_copy_texture = new D3DTexture2D(depth_texture.Get(), (D3D11_BIND_FLAG)depth_desc.BindFlags, depth_desc.Format);
+	return true;
+}
+
+bool D3DPostProcessor::ResizeStereoBuffer(const TargetSize& size)
+{
+	if (g_ActiveConfig.iStereoMode == STEREO_3DVISION)
+	{
+		TargetSize real_size(size.width * 2, size.height);
+		if (m_stereo_buffer_size == real_size)
+			return true;
+
+		// Create a staging texture for 3D vision with signature information in the last row.
+		// Nvidia 3D Vision supports full SBS, so there is no loss in resolution during this process.
+		int upload_buffer_size = (size.height + 1) * 4 * size.width * 2;
+		std::unique_ptr<u8[]> upload_buffer = std::make_unique<u8[]>(upload_buffer_size);
+		memset(upload_buffer.get(), 0, upload_buffer_size);
```

Comment:
upload buffer should be a vector in this case. It'll zero out its contents when given a size.